app/src/main/java/bacon/john/bitcoinpricewatch/bitcoin/BitcoinCurrencyManager.java

```diff
@@ -0,0 +1,120 @@
+                        public void execute(Realm realm) {
+                            Iterator<String> iter = response.keys();
+                            while (iter.hasNext()) {
+                                String code = iter.next();
+                                // filter out "all" since it really is not a ticker
+                                if (code.equals("all")) {
+                                    continue;
+                                }
+                                BitcoinCurrencyModel model = realm.where(BitcoinCurrencyModel.class).
+                                        equalTo("code", code).findFirst();
+                                if (model == null) {
+                                    model = new BitcoinCurrencyModel();
+                                    model.setCode(code);
+                                    realm.copyToRealm(model);
+                                }
+                            }
+                        }
+                    }, new Realm.Transaction.Callback() {
+                        @Override
+                        public void onSuccess() {
+                            callback.callback(BitcoinCurrencyManager.sharedInstance().getCachedCurrencies(), null);
+                        }
+                    });
+                }
+            }
+        });
+    }
+
+    public void getRemoteCurrencyPrices(BitcoinCurrencyModel currency, final CurrencyCallback callback) {
+        final String code = currency.getCode();
+        BitcoinCurrencyService.sharedInstance().fetchCurrencyPrices(code, new BitcoinCurrencyService.Callback() {
+            @Override
+            public void callback(final JSONObject response, Exception error) {
+                if (error != null) {
+                    callback.callback(null, error);
+                    return;
+                } else {
+                    LocalStorageService.sharedInstance().getRealm().executeTransaction(new Realm.Transaction() {
+                        @Override
+                        public void execute(Realm realm) {
+                            BitcoinCurrencyModel model = realm.where(BitcoinCurrencyModel.class).
+                                    equalTo("code", code).findFirst();
+
+                            try {
+                                model.setAsk(response.getDouble("ask"));
+                                model.setBid(response.getDouble("bid"));
+                                model.setLast(response.getDouble("last"));
+
+                                String updatedAt = response.getString("timestamp");
+                                DateFormat format = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z", Locale.ENGLISH);
+                                Date date = format.parse(updatedAt);
+                                model.setUpdatedAt(date.getTime());
+                            } catch (Exception e) {
+                                callback.callback(null, e);
```

Comment:
I'm aware that this callback is fired on a non-main thread, but really it ought to be; might be unexpected from the consumer's POV; wasn't sure what the best way to do this was from so deep in the parsing code